pkg/search/controller.go

```diff
@@ -0,0 +1,395 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+type clusterRegistry struct {
+	registries map[string]struct{}
+	resources  map[schema.GroupVersionResource]struct{}
+}
+
+func (c *clusterRegistry) unregistry() bool {
+	return len(c.registries) == 0
+}
+
+// Controller ResourceRegistry controller
+type Controller struct {
+	restConfig      *rest.Config
+	restMapper      meta.RESTMapper
+	informerFactory informerfactory.SharedInformerFactory
+	clusterLister   clusterlister.ClusterLister
+	queue           workqueue.RateLimitingInterface
+
+	clusterRegistry sync.Map
+
+	resourceHandler cache.ResourceEventHandler
+	InformerManager informermanager.MultiClusterInformerManager
+}
+
+// NewController returns a new ResourceRegistry controller
+func NewController(restConfig *rest.Config, rh cache.ResourceEventHandler) (*Controller, error) {
+	karmadaClient := karmadaclientset.NewForConfigOrDie(restConfig)
+	factory := informerfactory.NewSharedInformerFactory(karmadaClient, 0)
+	clusterLister := factory.Cluster().V1alpha1().Clusters().Lister()
+	restMapper, err := apiutil.NewDynamicRESTMapper(restConfig)
+	if err != nil {
+		klog.Errorf("Failed to create REST mapper: %v", err)
+		return nil, err
+	}
+
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	c := &Controller{
+		restConfig:      restConfig,
+		informerFactory: factory,
+		clusterLister:   clusterLister,
+		queue:           queue,
+		restMapper:      restMapper,
+
+		resourceHandler: rh,
+		InformerManager: informermanager.GetInstance(),
+	}
+	c.addAllEventHandlers()
+	return c, nil
+}
+
+// addAllEventHandlers adds all event handlers to the informer
+func (c *Controller) addAllEventHandlers() {
+	clusterInformer := c.informerFactory.Cluster().V1alpha1().Clusters().Informer()
+	clusterInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCluster,
+		UpdateFunc: c.updateCluster,
+		DeleteFunc: c.deleteCluster,
+	})
+
+	resourceRegistryInformer := c.informerFactory.Search().V1alpha1().ResourceRegistries().Informer()
+	resourceRegistryInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addResourceRegistry,
+		UpdateFunc: c.updateResourceRegistry,
+		DeleteFunc: c.deleteResourceRegistry,
+	})
+}
+
+// Start the controller
+func (c *Controller) Start(stopCh <-chan struct{}) {
+	klog.Infof("Starting karmada search controller")
+
+	defer runtime.HandleCrash()
+
+	c.informerFactory.Start(stopCh)
+	c.informerFactory.WaitForCacheSync(stopCh)
+
+	go wait.Until(c.worker, time.Second, stopCh)
+
+	go func() {
+		<-stopCh
+		informermanager.StopInstance()
+		klog.Infof("Shutting down karmada search controller")
+	}()
+}
+
+// worker processes the queue of resourceRegistry objects.
+func (c *Controller) worker() {
+	for c.cacheNext() {
+	}
+}
+
+// cacheNext processes the next resourceRegistry object in the queue.
```

Comment:
I guess you forget to update the comments here. It is `cluster` in the queue, not `resourceRegistry`.